File: apps/mobile/lib/core/logic/fabrary_import_apply.dart
import '../models/binder.dart';
import '../models/binder_entry.dart';
import 'fabrary_csv.dart';
import 'fabrary_match.dart';
import 'free_limits.dart';

class FabraryAdd {
  const FabraryAdd({required this.printingId, required this.quantity});

  final String printingId;
  final int quantity;

  Map<String, dynamic> toJson() => {
        'printingId': printingId,
        'quantity': quantity,
      };
}

class FabraryImportPlan {
  const FabraryImportPlan({
    required this.ok,
    this.refuseReason,
    required this.ownedCount,
    required this.matchedCount,
    required this.copiesToAdd,
    required this.unmatched,
    required this.adds,
    this.binderId,
  });

  final bool ok;
  final String? refuseReason;
  final int ownedCount;
  final int matchedCount;
  final int copiesToAdd;
  final List<FabraryUnmatched> unmatched;
  final List<FabraryAdd> adds;
  final String? binderId;

  Map<String, dynamic> toJson() => {
        'ok': ok,
        'refuseReason': refuseReason,
        'ownedCount': ownedCount,
        'matchedCount': matchedCount,
        'copiesToAdd': copiesToAdd,
        'unmatched': unmatched.map((u) => u.toJson()).toList(),
        'adds': adds.map((a) => a.toJson()).toList(),
      };
}

class _HaveParse {
  const _HaveParse._(this.kind, this.quantity);
  const _HaveParse.empty() : this._('empty', 0);
  const _HaveParse.invalid() : this._('invalid', 0);
  const _HaveParse.owned(int qty) : this._('owned', qty);

  final String kind;
  final int quantity;
}

_HaveParse parseHave(dynamic raw) {
  final text = '$raw'.trim();
  if (text.isEmpty) return const _HaveParse.empty();
  final n = num.tryParse(text);
  if (n == null || !n.isFinite) return const _HaveParse.invalid();
  if (n <= 0) return const _HaveParse.empty();
  return _HaveParse.owned(n.toInt());
}

List<String> ownedIdsFromEntries(Iterable<dynamic> entries) {
  final ids = <String>{};
  for (final entry in entries) {
    if (entry is BinderEntry) {
      if (entry.isWanted) continue;
      final id = entry.card.id;
      if (id.isNotEmpty) ids.add(id);
      continue;
    }
    if (entry is Map) {
      if (entry['isWanted'] == true) continue;
      final id = '${entry['printingId'] ?? entry['cardId'] ?? ''}';
      if (id.isNotEmpty) ids.add(id);
    }
  }
  return ids.toList();
}

FabraryImportPlan _empty(String reason) => FabraryImportPlan(
      ok: false,
      refuseReason: reason,
      ownedCount: 0,
      matchedCount: 0,
      copiesToAdd: 0,
      unmatched: const [],
      adds: const [],
    );

FabraryImportPlan planFabraryImport({
  List<String>? headers,
  List<Map<String, dynamic>>? rows,
  String? csv,
  required List<dynamic> catalog,
  required String binderId,
  List<dynamic> existingEntries = const [],
  List<String>? existingOwnedIds,
  bool isPro = false,
  bool notFabrary = false,
}) {
  var resolvedHeaders = headers;
  var resolvedRows = rows;
  if (csv != null) {
    final parsed = parseFabraryCsv(csv);
    if (!parsed.ok) return _empty('not_fabrary');
    resolvedHeaders = parsed.headers;
    resolvedRows = parsed.rows;
  }
  if (notFabrary || !hasFabraryHeaders(resolvedHeaders)) {
    return _empty('not_fabrary');
  }

  final unmatched = <FabraryUnmatched>[];
  final qtyById = <String, int>{};
  var ownedCount = 0;

  for (final row in resolvedRows ?? const <Map<String, dynamic>>[]) {
    final have = parseHave(row['Have']);
    if (have.kind == 'empty') continue;
    if (have.kind == 'invalid') {
      ownedCount += 1;
      unmatched.add(
        FabraryUnmatched(
          name: '${row['Name'] ?? ''}',
          setNumber: '${row['Set number'] ?? ''}',
          foiling: '${row['Foiling'] ?? ''}',
          treatment: '${row['Treatment'] ?? ''}',
          edition: '${row['Edition'] ?? ''}',
        ),
      );
      continue;
    }
    ownedCount += 1;
    final match = matchFabraryRow(row, catalog);
    if (match.unmatched != null) {
      unmatched.add(match.unmatched!);
      continue;
    }
    final id = match.printingId!;
    qtyById[id] = (qtyById[id] ?? 0) + have.quantity;
  }

  if (ownedCount == 0) return _empty('no_owned');

  final adds = [
    for (final entry in qtyById.entries)
      FabraryAdd(printingId: entry.key, quantity: entry.value),
  ];
  final matchedCount = adds.length;
  final copiesToAdd = adds.fold<int>(0, (s, a) => s + a.quantity);

  if (matchedCount == 0) {
    return FabraryImportPlan(
      ok: false,
      refuseReason: 'no_matched',
      ownedCount: ownedCount,
      matchedCount: 0,
      copiesToAdd: 0,
      unmatched: unmatched,
      adds: const [],
      binderId: binderId,
    );
  }

  final existingIds = existingOwnedIds ?? ownedIdsFromEntries(existingEntries);
  final incomingIds = [for (final add in adds) add.printingId];
  if (!FreeLimits.canImportDistinctPrintings(
    existingIds,
    incomingIds,
    isPro: isPro,
  )) {
    return FabraryImportPlan(
      ok: false,
      refuseReason: 'free_cap',
      ownedCount: ownedCount,
      matchedCount: matchedCount,
      copiesToAdd: 0,
      unmatched: unmatched,
      adds: const [],
      binderId: binderId,
    );
  }

  return FabraryImportPlan(
    ok: true,
    refuseReason: null,
    ownedCount: ownedCount,
    matchedCount: matchedCount,
    copiesToAdd: copiesToAdd,
    unmatched: unmatched,
    adds: adds,
    binderId: binderId,
  );
}

/// Combine Have quantities onto existing NM rows in this Binder only.
List<Map<String, dynamic>> applyImportAddsToEntries(
  List<Map<String, dynamic>> existingEntries,
  String binderId,
  List<FabraryAdd> adds,
) {
  final next = [
    for (final entry in existingEntries) Map<String, dynamic>.from(entry),
  ];
  for (final add in adds) {
    final idx = next.indexWhere((entry) {
      if (entry['isWanted'] == true) return false;
      final entryBinder = '${entry['binderId'] ?? BinderIds.trade}';
      final condition = '${entry['condition'] ?? 'NM'}';
      final id = '${entry['printingId'] ?? entry['cardId'] ?? ''}';
      return entryBinder == binderId && condition == 'NM' && id == add.printingId;
    });
    if (idx >= 0) {
      final qty = (next[idx]['quantity'] as num?)?.toInt() ?? 0;
      next[idx]['quantity'] = qty + add.quantity;
    } else {
      next.add({
        'printingId': add.printingId,
        'cardId': add.printingId,
        'binderId': binderId,
        'isWanted': false,
        'quantity': add.quantity,
        'condition': 'NM',
      });
    }
  }
  return next;
}

File: apps/web/src/utils/fabraryImportApply.js
/**
 * Plan a Fabrary import:
 * Have → Collection, Want in trade / Want to buy → Want List,
 * Extra for trade / Extra to sell → Trade Binder. Add-on-top. No free-tier cap.
 */

import { parseFabraryCsv, hasFabraryHeaders } from './fabraryCsv.js';
import { buildSetCodeIndex, matchFabraryRow } from './fabraryMatch.js';

export const FABRARY_DESTINATION = {
    collection: 'collection',
    trade: 'trade',
    want: 'want',
};

export function parseQty(raw) {
    const text = String(raw ?? '').trim();
    if (text === '') return { kind: 'empty', quantity: 0 };
    const n = Number(text);
    if (!Number.isFinite(n)) return { kind: 'invalid', quantity: 0 };
    if (n <= 0) return { kind: 'empty', quantity: 0 };
    return { kind: 'owned', quantity: n };
}

/** @deprecated Use parseQty. Have-only alias kept for existing callers. */
export function parseHave(raw) {
    return parseQty(raw);
}

function qtyOf(row, ...keys) {
    let quantity = 0;
    let invalid = false;
    let seen = false;
    for (const key of keys) {
        const parsed = parseQty(row?.[key]);
        if (parsed.kind === 'invalid') invalid = true;
        if (parsed.kind === 'owned') {
            seen = true;
            quantity += parsed.quantity;
        }
    }
    return { quantity, invalid, seen };
}

function unmatchedFromRow(row) {
    return {
        name: String(row?.Name ?? ''),
        setNumber: String(row?.['Set number'] ?? ''),
        foiling: String(row?.Foiling ?? ''),
        treatment: String(row?.Treatment ?? ''),
        edition: String(row?.Edition ?? ''),
    };
}

function emptyPlan(refuseReason, extras = {}) {
    return {
        ok: false,
        refuseReason,
        ownedCount: 0,
        matchedCount: 0,
        copiesToAdd: 0,
        unmatched: [],
        adds: [],
        ...extras,
    };
}

function addQty(map, printingId, quantity) {
    map.set(printingId, (map.get(printingId) || 0) + quantity);
}

function addsFromMap(map, destination) {
    return [...map.entries()].map(([printingId, quantity]) => ({
        printingId,
        quantity,
        destination,
    }));
}

export function copiesForDestination(adds, destination) {
    return (adds || [])
        .filter((add) => add.destination === destination)
        .reduce((sum, add) => sum + add.quantity, 0);
}

/**
 * @param {Object} input
 * @param {string[]} [input.headers]
 * @param {Object[]} [input.rows]
 * @param {string} [input.csv]
 * @param {Object[]} input.catalog
 */
export function planFabraryImport(input = {}) {
    let headers = input.headers;
    let rows = input.rows;
    if (input.csv != null) {
        const parsed = parseFabraryCsv(input.csv);
        if (!parsed.ok) return emptyPlan('not_fabrary');
        headers = parsed.headers;
        rows = parsed.rows;
    }
    if (input.notFabrary || !hasFabraryHeaders(headers)) {
        return emptyPlan('not_fabrary');
    }

    const unmatched = [];
    const collectionById = new Map();
    const tradeById = new Map();
    const wantById = new Map();
    let ownedCount = 0;
    const catalogIndex = buildSetCodeIndex(input.catalog || []);

    for (const row of rows || []) {
        const have = qtyOf(row, 'Have');
        const want = qtyOf(row, 'Want in trade', 'Want to buy');
        const extra = qtyOf(row, 'Extra for trade', 'Extra to sell');
        const hasValid = have.quantity > 0 || want.quantity > 0 || extra.quantity > 0;
        const hasInvalidOnly = !hasValid && (have.invalid || want.invalid || extra.invalid);
        if (!hasValid && !hasInvalidOnly) continue;

        ownedCount += 1;
        if (hasInvalidOnly) {
            unmatched.push(unmatchedFromRow(row));
            continue;
        }

        const match = matchFabraryRow(row, catalogIndex);
        if (match.unmatched) {
            unmatched.push(match.unmatched);
            continue;
        }
        if (have.quantity > 0) addQty(collectionById, match.printingId, have.quantity);
        if (want.quantity > 0) addQty(wantById, match.printingId, want.quantity);
        if (extra.quantity > 0) addQty(tradeById, match.printingId, extra.quantity);
    }

    if (ownedCount === 0) {
        return emptyPlan('no_owned');
    }

    const adds = [
        ...addsFromMap(collectionById, FABRARY_DESTINATION.collection),
        ...addsFromMap(wantById, FABRARY_DESTINATION.want),
        ...addsFromMap(tradeById, FABRARY_DESTINATION.trade),
    ];
    const matchedCount = adds.length;
    const copiesToAdd = adds.reduce((sum, add) => sum + add.quantity, 0);

    if (matchedCount === 0) {
        return {
            ok: false,
            refuseReason: 'no_matched',
            ownedCount,
            matchedCount: 0,
            copiesToAdd: 0,
            unmatched,
            adds: [],
        };
    }

    return {
        ok: true,
        refuseReason: null,
        ownedCount,
        matchedCount,
        copiesToAdd,
        unmatched,
        adds,
        restoreCollection: collectionById.size > 0,
    };
}

function binderIdFor(destination, collectionBinderId, tradeBinderId) {
    if (destination === FABRARY_DESTINATION.trade) return tradeBinderId;
    if (destination === FABRARY_DESTINATION.want) return null;
    return collectionBinderId;
}

/**
 * Combine planned quantities onto existing NM rows (or Want List rows).
 *
 * @param {Object[]} existingEntries
 * @param {{ printingId: string, quantity: number, destination?: string }[]} adds
 * @param {{ collectionBinderId?: string, tradeBinderId?: string }} [ids]
 * @returns {Object[]}
 */
export function applyImportAddsToEntries(
    existingEntries,
    adds,
    ids = {},
) {
    const collectionBinderId = ids.collectionBinderId || 'system:collection';
    const tradeBinderId = ids.tradeBinderId || 'system:trade';
    const next = (existingEntries || []).map((entry) => ({ ...entry }));
    for (const add of adds || []) {
        const destination = add.destination || FABRARY_DESTINATION.collection;
        const wanted = destination === FABRARY_DESTINATION.want;
        const binderId = binderIdFor(destination, collectionBinderId, tradeBinderId);
        const idx = next.findIndex((entry) => {
            const id = entry.printingId || entry.cardId;
            if (id !== add.printingId) return false;
            if (wanted) return Boolean(entry.isWanted);
            if (entry.isWanted) return false;
            return (entry.binderId || 'system:trade') === binderId
                && (entry.condition || 'NM') === 'NM';
        });
        if (idx >= 0) {
            next[idx] = {
                ...next[idx],
                quantity: (Number(next[idx].quantity) || 0) + add.quantity,
            };
        } else {
            next.push({
                printingId: add.printingId,
                cardId: add.printingId,
                binderId: wanted ? null : binderId,
                isWanted: wanted,
                quantity: add.quantity,
                condition: 'NM',
            });
        }
    }
    return next;
}
